### DISCUSSION TITLE
Testing cache updaters (for subscriptions)?

### DISCUSSION THREAD

**Discussion by bartenra** [urql-graphql/urql#1774] [Q&A] [answered]
There doesn't seem to be anything in the docs about this specifically, but any ideas how to best test cache updaters?

I'm trying something along the lines of the code below.
However, the results are very flaky. Sometimes the subscription result is printed and the cache is updated, sometimes the sleep time is not enough.

```ts
pipe(
  visitorClient.subscription(VisitorChatSubscriptionDocument),
  subscribe(result => {
    console.error("result.data", result?.data);
  })
);

await sleep(1000);

const { data: sendMessageToUserMutData, error } = await orgUserClient.mutation(SendMessageToUserDocument, {
  userId: accountRegister.user?.id,
  message: {
    content,
    type
  }
}).toPromise();
expect(sendMessageToUserMutData?.sendMessageToUser.error).toBeFalsy();
expect(error).toBeFalsy();

await sleep(1000);

const { data } = await visitorClient.query(VisitorChatDocument, {}, { requestPolicy: 'cache-only' }).toPromise();
expect(data?.me?.chats).toHaveLength(1);
```

**Comment by kitten** [comment, marked as answer]
We would indeed recommend more complete integration tests, like what you're presumably doing, where the client isn't fully mocked out but you're explicitly testing its interactions and updates 

I don't know whether you're actually mocking responses here though (our storybook add-on has an example of that kind of exchange) but the important part is that subscriptions aren't imperative, so they can't be tested as such.

If you need to test the timings of subscriptions then it's important to get those timings of subscription results right. For instance, I can see that you have a query here but it only runs after all of the other assertions. You have to both ensure that the query is initialised and that the subscription may update it.

There are certain timing and limiting utilities in Wonka that can be useful for this purpose. For instance, you can `take` a limited amount of subscription results and then use `toPromise` to wait for that exact amount.

In general, we've got some examples of these kinds of tests in our own tests in `@urql/core` and `@urql/exchange-graphcache`. With some more details I can probably paste some pseudo code here

**Reply by bartenra** [reply]
Correct. It's an intergration test. The server is running locally.

`take` looks more elegant than what I was doing now:

```ts
let messagesReceived = 0;

const { unsubscribe } = pipe(
  orgUserClient.subscription(OrganizationChatSubscriptionDocument),
  subscribe(result => {
    ++messagesReceived;
  }),
);

await orgUserClient.query(OrganizationUserChatsDocument).toPromise();

// do mutation
...

while (messagesReceived != 2) {
  await sleep(300);
}
// check if cache is updated correctly
...
unsubscribe()
```

But I am not sure how to use `take` correctly. I am trying this, but the promise never resolves:


```ts
// initialize cache with empty data, so the cache updater can add a chat to the pre-existing empty array in the cache
await orgUserClient.query(OrganizationUserChatsDocument).toPromise();

const subcriptionPromise = pipe(
  orgUserClient.subscription(OrganizationChatSubscriptionDocument),
  take(2),
  toPromise
);

 // do mutation
 ...

await subcriptionPromise;  // wait for messages to be received and processed by cache updater

// check if cache is updated correctly
...
```

**Reply by kitten** [reply]
Have you tried slicing these calls up and seeing what exactly is going on? You could use `onPush` to see whether you're even getting the results and the amount you're expecting, or drop the `debugExchange` into your client to see whether it's seeing all the results you need

**Reply by bartenra** [reply]
Ah, the `debugExchange` was indeed very helpful.

Thanks a bunch.
